include/evmc.h

```diff
@@ -140,51 +140,105 @@ typedef void (*evmc_get_block_hash_fn)(struct evmc_uint256be* result,
                                        struct evmc_context* context,
                                        int64_t number);
 
-/// The execution status code.
-enum evmc_status_code {
-    EVMC_SUCCESS = 0,               ///< Execution finished with success.
-    EVMC_FAILURE = 1,               ///< Generic execution failure.
-    EVMC_OUT_OF_GAS = 2,
-    EVMC_UNDEFINED_INSTRUCTION = 3, ///< Unknown instruction encountered by the VM.
-    EVMC_BAD_JUMP_DESTINATION = 4,
-    EVMC_STACK_OVERFLOW = 5,
-    EVMC_STACK_UNDERFLOW = 6,
-    EVMC_REVERT = 7,                ///< Execution terminated with REVERT opcode.
-
-    /// Tried to execute an operation which is restricted in static mode.
-    EVMC_STATIC_MODE_VIOLATION = 8,
-
-    /// The dedicated INVALID instruction was hit.
-    EVMC_INVALID_INSTRUCTION = 9,
-
-    /// Tried to read outside memory bounds.
-    ///
-    /// An example is RETURNDATACOPY reading past the available buffer.
-    EVMC_INVALID_MEMORY_ACCESS = 10,
-
-    /// Exceptions produced by precompiles/system contracts
-    ///
-    /// An example: elliptic curve functions handed invalid EC points
-    EVMC_PRECOMPILE_FAILURE = 11,
-
-    /// Call depth exceded (if there is a call depth limit)
-    EVMC_CALL_DEPTH_EXCEDED = 12,
-
-    /// The EVM rejected the execution of the given code or message.
-    ///
-    /// This error SHOULD be used to signal that the EVM is not able to or
-    /// willing to execute the given code type or message.
-    /// If an EVM returns the ::EVMC_REJECTED status code,
-    /// the Client MAY try to execute it in other EVM implementation.
-    /// For example, the Client tries running a code in the EVM 1.5. If the
-    /// code is not supported there, the execution falls back to the EVM 1.0.
-    EVMC_REJECTED = -1,
-
-    /// EVM implementation internal error.
-    ///
-    /// @todo We should rethink reporting internal errors. One of the options
-    ///       it to allow using any negative value to represent internal errors.
-    EVMC_INTERNAL_ERROR = -2,
+/**
+ * The execution status code.
+ *
+ * Successful execution is represented by ::EVMC_SUCCESS having value 0.
+ *
+ * Positive values represent failures defined by VM specifications with generic
+ * ::EVMC_FAILURE code of value 1.
+ *
+ * Status codes with negative values represent VM internal errors
+ * not provided by EVM specifications. These errors MUST not be passed back
+ * to the caller. They MAY be handled by the Client in predefined manner
+ * (see e.g. ::EVMC_REJECTED), otherwise internal errors are not recoverable.
+ * The generic representant of errors is ::EVMC_INTERNAL_ERROR but
+ * an EVM implementation MAY return negative status codes that are not defined
+ * in the EVMC documentation.
```

Comment:
Should probably leave a comment here along the lines of "should you need more error codes, please create an issue on the evmc repo to discuss" to facilitate discussion.

Reply:
Added.